openpnp/openpnp
# Issue #329 - fiducial pipeline

This is a first draft of an enhancement to allow editing of a fiducial part pipeline.  It stores a pipeline XML config in the Part Alignment section of machine.xml.

The only change to the GUI is a checkbox on the (Part Tab) -> (Pick any part) lower panel "Alignment", which makes the existing "Edit" button use the default (top) camera instead of the bottom vision camera.

The fiducial locator logic now uses that pipeline from the bottom vision config data to find keypoints under "results" of processing the pipeline and uses the closest one.  This needs to be upgraded to using a template match, but that will require a little more research by me.  This is a demo to see if everyanybody want to continue in this direction.

How to test this:
1. Start machine and Home
2. Load a job with a board with a fiducial, put board on table
3. move camera to fiducial
4. In "Part" tab, find the fiducial part.
5. Note new checkbox at bottom of Alignment panel.  Try Editing the pipeline with the box checked/unchecked.  The only difference is the dialog name and the camera used, the XML data is per part.
6. set up a pipeline which returns keypoints named "results" and tags the fiducial well
7. go to the "Jobs" tab and click the magic button for a fiducial check and position update.

The error I always get at this point (if the pipeline is working) is something about outside 1% of expected, so the pipeline has been processed for the two most distant fiducials.  That's all I want to do for this first draft, as I'm looking to make this work right, and I'm very very new around here.

Okay, so what questions need to be asked here? 
* I don't think this is what somebody originally had in mind, but per part pipeline is already supported by the rest of the code here.  Is this acceptable?  Will this change or prevent any planned model schema changes?  Should this be tweaked, scrapped, or is it almost good to go?
* While parts have a top and bottom, fiducials are virtual and only have a shadow.  There's no need to store a bottom vision pipeline for a fiducial, but I can see why a separate top vision pipeline for a part might be useful.  This isn't a fixed-length storage system, and this can be added if needed to the XML as an additional pipeline (or an array, if a pipeline per part per camera).  
* The checkbox is automatically populated if the lower-case part ID starts with "fiducial", but is otherwise manually operated (not persistent). How's the wording on that?  Does it take you less than 15 seconds to figure out what it does just based on it's appearance?

I have to figure out how to do the template match thing, not expecting that to be too hard.

--visualecho1
